system-addon/lib/MessageManager.jsm

```diff
@@ -0,0 +1,159 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+/* globals AboutNewTab, uuidgen, RemotePages, XPCOMUtils */
+"use strict";
+
+const {utils: Cu} = Components;
+
+Cu.import("resource://gre/modules/RemotePageManager.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+const {
+  actionUtils: au,
+  actionCreators: ac,
+  actionTypes: at
+} = Cu.import("resource://activity-stream/common/Actions.jsm", {});
+
+XPCOMUtils.defineLazyModuleGetter(this, "AboutNewTab",
+  "resource:///modules/AboutNewTab.jsm");
+
+const ABOUT_NEW_TAB_URL = "about:newtab";
+
+const DEFAULT_OPTIONS = {
+  dispatch(action) {dump(`\nMessage manager: Received action ${action.type}, but no dispatcher was defined.\n`)},
+  pageURL: ABOUT_NEW_TAB_URL,
+  outgoingMessageName: "ActivityStream:MainToContent",
+  incomingMessageName: "ActivityStream:ContentToMain"
+};
+
+class MessageManager {
+  constructor(options = {}) {
+    Object.keys(DEFAULT_OPTIONS).forEach(o => {
+      this[o] = typeof options[o] !== "undefined" ?  options[o] : DEFAULT_OPTIONS[o];
+    });
+    this.channel = null;
+
+    this.middleware = this.middleware.bind(this);
+    this.onMessage = this.onMessage.bind(this);
+    this.onNewTabLoad = this.onNewTabLoad.bind(this);
+    this.onNewTabUnload = this.onNewTabUnload.bind(this);
```

Comment:
I think I'm actually going to leave them like this, just because testing/comparing is easier, and down the road if we refactor we could accidentally forget to remove listeners